pkg/koordlet/runtimehooks/nri/server.go

```diff
@@ -0,0 +1,189 @@
+	updateContainer func(*NriServer, *api.PodSandbox, *api.Container) ([]*api.ContainerUpdate, error)
+}
+
+var (
+	_ = stub.ConfigureInterface(&NriServer{})
+	_ = stub.SynchronizeInterface(&NriServer{})
+	_ = stub.RunPodInterface(&NriServer{})
+	_ = stub.CreateContainerInterface(&NriServer{})
+	_ = stub.UpdateContainerInterface(&NriServer{})
+
+	pluginName = "koordlet_nri"
+	pluginIdx  = "00"
+	events     = "RunPodSandbox,CreateContainer,UpdateContainer"
+	cfg        nriconfig
+	opts       []stub.Option
+	err        error
+)
+
+func NewNriServer(opt Options) (*NriServer, error) {
+	if Enabled() {
+		opts = append(opts, stub.WithPluginName(pluginName))
+		opts = append(opts, stub.WithPluginIdx(pluginIdx))
+		p := &NriServer{options: opt}
+		if p.mask, err = api.ParseEventMask(events); err != nil {
+			klog.Errorf("failed to parse events: %v", err)
+		}
+		cfg.Events = strings.Split(events, ",")
+
+		if p.stub, err = stub.New(p, append(opts, stub.WithOnClose(p.onClose))...); err != nil {
+			klog.Errorf("failed to create plugin stub: %v", err)
+		}
+
+		return p, err
+	}
+	return nil, nil
+}
+
+func (s *NriServer) Start() error {
+	go func() {
+		klog.Info("-------nri start-----------")
+		err := s.stub.Run(context.Background())
+		if err != nil {
+			klog.Errorf("plugin exited with error %v", err)
+		}
+	}()
+	return nil
+}
+
+func Enabled() bool {
+	return true
+	//return features.DefaultKoordletFeatureGate.Enabled(features.NRIHooksManager)
+}
+
+func (p *NriServer) Configure(config, runtime, version string) (stub.EventMask, error) {
+	klog.Infof("got configuration data: %q from runtime %s %s", config, runtime, version)
+	if config == "" {
+		return p.mask, nil
+	}
+
+	err := yaml.Unmarshal([]byte(config), &cfg)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse provided configuration: %w", err)
+	}
+
+	p.mask, err = api.ParseEventMask(cfg.Events...)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse events in configuration: %w", err)
+	}
+
+	return p.mask, nil
+}
+
+func (p *NriServer) Synchronize(pods []*api.PodSandbox, containers []*api.Container) ([]*api.ContainerUpdate, error) {
+	return nil, nil
+}
+
+func (p *NriServer) RunPodSandbox(pod *api.PodSandbox) error {
+	podCtx := &protocol.PodContext{}
+	podCtx.FromNri(pod)
+	err := hooks.RunHooks(p.options.PluginFailurePolicy, rmconfig.PreRunPodSandbox, podCtx)
+	if err != nil {
+		klog.Errorf("hooks run error: %v", err)
+	}
+	podCtx.NriDone(p.options.Executor)
+	return nil
+}
+
+func (p *NriServer) CreateContainer(pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+	containerCtx := &protocol.ContainerContext{}
+	containerCtx.FromNri(pod, container)
+	err := hooks.RunHooks(p.options.PluginFailurePolicy, rmconfig.PreCreateContainer, containerCtx)
+	if err != nil {
+		klog.Errorf("run hooks error: %v", err)
```

Comment:
"return err" or "bypass the error"?

Reply:
should we just return error or the error will be passed to containerd ?